src/Data/LargeHashable/Class.hs

```diff
@@ -5,63 +5,94 @@
-import qualified Data.Text.Internal.Lazy as TLI
+import qualified Codec.Binary.UTF8.Light as Utf8
+import qualified Data.Aeson as J
 import qualified Data.ByteString as B
-import qualified Data.ByteString.Short as BS
 import qualified Data.ByteString.Lazy as BL
 import qualified Data.ByteString.Lazy.Internal as BLI
-import qualified Data.Set as S
-import qualified Data.IntSet as IntSet
+import qualified Data.ByteString.Short as BS
+import qualified Data.Foldable as F
+import qualified Data.HashMap.Lazy as HashMap
 import qualified Data.HashSet as HashSet
-import qualified Data.Map as M
 import qualified Data.IntMap as IntMap
-import qualified Data.HashMap.Lazy as HashMap
-import Data.Time
-import Data.Time.Clock.TAI
+import qualified Data.IntSet as IntSet
+import qualified Data.Map as M
+import qualified Data.Scientific as Sci
+import qualified Data.Sequence as Seq
+import qualified Data.Set as S
+import qualified Data.Strict.Tuple as Tuple
+import qualified Data.Text as T
+import qualified Data.Text.Foreign as TF
+import qualified Data.Text.Internal.Lazy as TLI
+import qualified Data.Text.Lazy as TL
+import qualified Data.Vector as V
 
 -- | A type class for computing hashes (i.e. MD5, SHA256, ...) from
 -- haskell values.
 --
--- Important: when implementing `LargeHashable` for your custom datatype,
--- make sure to hash all information present in the datatype. Otherwise,
--- unncessary hash collisions arise. A rule of thumb: hash all
--- information that you would also need for serializing/deserializing
--- values of your datatype.
+-- The laws of this typeclass are the following:
 --
--- The law of this typeclass is the following: If two values are equal
+-- (1) If two values are equal
 -- according to '==', then the finally computed hashes must also be equal
 -- according to '=='. However it is not required that the hashes of inequal
 -- values have to be inequal. Also note that an instance of 'LargeHashable'
 -- does not require a instance of 'Eq'. Using any sane algorithm the chance
 -- of a collision should be 1 / n where n is the number of different hashes
 -- possible.
+--
+-- (2) If two values are inequal
+-- according to '==', then the probability of a hash collision should
+-- be 1/n, where n is the number of possible hashes produced by the
+-- underlying hash algorithm.
+--
+-- A rule of thumb: hash all information that you would also need for
+-- serializing/deserializing values of your datatype. For instance, when
+-- hashing lists, you would not only hash the list elements but also the
+-- length of the list. Consider the following datatype
+--
+-- > data Foo = Foo [Int] [Int]
+--
+-- We now write an instance for LargeHashable like this
+--
+-- > instance LargeHashable Foo where
+-- >     updateHash (Foo l1 l2) = updateHash l1 >> updateHash l2
+--
+-- If we did not hash the length of a list, then the following two values
+-- of @Foo@ would produce identical hashes:
+--
+-- > Foo [1,2,3] []
+-- > Foo [1] [2,3]
```

Comment:
Yes, but then you would need to think about the delimiter when writing the instance for the datatype containing the list. The idea is that you have to think only about the properties of the type you are writing the instance for.